### PR TITLE
domain,infoschema: fix the co-exist of infoschema v1 and v2 causing memory leak

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: ref #50959 

Problem Summary:

When tidb start up, it's using infoschema v1 because DefTiDBSchemaCache is v1.
And after it load the @@global.tidb_schema_cache_size variable and found it's non-zero, infoschema v2 is used.

Then infoschema v1 and v2 co-exist cause the memory usage doubled.

### What changed and how does it work?

1. In info cache this commit replace the old v1 with the new v2.
2. The schema validator is reference the old v1 cause it can not be GC and freed, remove that referencee
3. Somewhere else seems also reference the old v1, but it's too difficult to find out where, using 'let it crash' strategy to figure out later.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Before infoschema v1 been GC, both v1 and v2 exists: 

image

This PR after the reset of v1 object:

image

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by tiancaiamao** [comment]
A data race on current master branch, not caused by this commit, so PTAL @wjhuang2016 @ywqzzy @GMHDBJD 

```
==================
WARNING: DATA RACE
Read at 0x00c0053e6008 by goroutine 304:
  github.com/pingcap/tidb/pkg/session.(*LazyTxn).Valid()
      pkg/session/txn.go:227 +0x39
  github.com/pingcap/tidb/pkg/testkit.(*MockSessionManager).GetInternalSessionStartTSList()
      pkg/testkit/mocksessionmanager.go:164 +0x2e2
  github.com/pingcap/tidb/pkg/domain/infosync.(*InfoSyncer).ReportMinStartTS()
      pkg/domain/infosync/info.go:756 +0xb9
  github.com/pingcap/tidb/pkg/domain.(*Domain).infoSyncerKeeper()
      pkg/domain/domain.go:793 +0x53a
  github.com/pingcap/tidb/pkg/domain.(*Domain).infoSyncerKeeper-fm()
      <autogenerated>:1 +0x33
  github.com/pingcap/tidb/pkg/util.(*WaitGroupEnhancedWrapper).Run.func1()
      pkg/util/wait_group_wrapper.go:99 +0xf4

Previous write at 0x00c0053e6008 by goroutine 765414:
  github.com/pingcap/tidb/pkg/session.(*LazyTxn).changeToPending()
      pkg/session/txn.go:284 +0x524
  github.com/pingcap/tidb/pkg/session.(*session).PrepareTSFuture()
      pkg/session/session.go:3924 +0x388
  github.com/pingcap/tidb/pkg/sessiontxn/isolation.(*baseTxnContextProvider).replaceTxnTsFuture()
      pkg/sessiontxn/isolation/base.go:367 +0x1c7
  github.com/pingcap/tidb/pkg/sessiontxn/isolation.(*baseTxnContextProvider).prepareTxn()
      pkg/sessiontxn/isolation/base.go:332 +0x217
  github.com/pingcap/tidb/pkg/sessiontxn/isolation.(*baseTxnContextProvider).AdviseWarmup()
      pkg/sessiontxn/isolation/base.go:392 +0x97
  github.com/pingcap/tidb/pkg/sessiontxn/isolation.(*OptimisticTxnContextProvider).AdviseWarmup()
      <autogenerated>:1 +0x31
  github.com/pingcap/tidb/pkg/session.(*txnManager).AdviseWarmup()
      pkg/session/txnmanager.go:317 +0xe6
  github.com/pingcap/tidb/pkg/planner/contextimpl.(*PlanCtxExtendedImpl).AdviseTxnWarmup()
      pkg/planner/contextimpl/impl.go:55 +0x85
  github.com/pingcap/tidb/pkg/session.(*planContextImpl).AdviseTxnWarmup()
      <autogenerated>:1 +0x46
  github.com/pingcap/tidb/pkg/planner.Optimize()
      pkg/planner/optimize.go:230 +0x1025
  github.com/pingcap/tidb/pkg/executor.(*Compiler).Compile()
      pkg/executor/compiler.go:99 +0x884
  github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt()
      pkg/session/session.go:2105 +0x11dd
  github.com/pingcap/tidb/pkg/session.(*session).ExecuteInternal()
      pkg/session/session.go:1530 +0x3f7
  github.com/pingcap/tidb/pkg/util/sqlexec.ExecSQL()
      pkg/util/sqlexec/restricted_sql_executor.go:252 +0xee
  github.com/pingcap/tidb/pkg/disttask/framework/storage.(*TaskManager).ExecuteSQLWithNewSession.func1()
      pkg/disttask/framework/storage/task_table.go:183 +0xf8
  github.com/pingcap/tidb/pkg/disttask/framework/storage.(*TaskManager).WithNewSession()
      pkg/disttask/framework/storage/task_table.go:147 +0x1a1
  github.com/pingcap/tidb/pkg/disttask/framework/storage.(*TaskManager).ExecuteSQLWithNewSession()
      pkg/disttask/framework/storage/task_table.go:182 +0x184
  github.com/pingcap/tidb/pkg/disttask/framework/storage.(*TaskManager).GetSubtaskCntGroupByStates()
      pkg/disttask/framework/storage/task_table.go:507 +0x167
  github.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*BaseScheduler).onRunning()
      pkg/disttask/framework/scheduler/scheduler.go:374 +0x39d
  github.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*BaseScheduler).scheduleTask()
      pkg/disttask/framework/scheduler/scheduler.go:255 +0x1504
  github.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*BaseScheduler).ScheduleTask()
      pkg/disttask/framework/scheduler/scheduler.go:119 +0x268
  github.com/pingcap/tidb/pkg/disttask/importinto.(*importScheduler).ScheduleTask()
      <autogenerated>:1 +0x3e
  github.com/pingcap/tidb/pkg/disttask/framework/scheduler.(*Manager).startScheduler.func1()
      pkg/disttask/framework/scheduler/scheduler_manager.go:367 +0x1e1
  github.com/pingcap/tidb/pkg/util.(*WaitGroupWrapper).RunWithLog.func1()
      pkg/util/wait_group_wrapper.go:171 +0xf1
```

File: pkg/domain/domain.go [hunk 1]
```diff
@@ -319,18 +360,27 @@ func (do *Domain) loadInfoSchema(startTS uint64) (infoschema.InfoSchema, bool, i
...
+		zap.Bool("isV2", isV2),
 		zap.Int64("currentSchemaVersion", currentSchemaVersion),
 		zap.Int64("neededSchemaVersion", neededSchemaVersion),
-		zap.Duration("start time", time.Since(startTime)))
+		zap.Duration("elapsed time", time.Since(startTime)))
 
-	is := newISBuilder.Build(startTS)
-	do.infoCache.Insert(is, schemaTs)
+	if isV1V2Switch && schemaTs > 0 {
+		// Reset the whole info cache to avoid co-existing of both v1 and v2, causing the memory usage doubled.
```
Comment (wjhuang2016) [inline pkg/domain/domain.go:378]: So it may take 10 mins to free the v1 memory?

Comment (tiancaiamao) [reply pkg/domain/domain.go:378]: Yes, if we reset it immediately, the bootstrap background jobs may get panic.

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Review by ywqzzy — APPROVED** [review]

**Comment by tiancaiamao** [comment]
I create a issue for the DATA RACE https://github.com/pingcap/tidb/issues/54513

**Comment by tiancaiamao** [comment]
Unstable CI https://github.com/pingcap/tidb/issues/54516  but it's not caused by the changes in this PR

**Pull request merged**